### PR TITLE
Being able to edit project update using pencil edit icon

### PR DESCRIPTION
**Pull request by mrpatrickwest**
Was throwing an exception that editConfiguration not returned.

### CODE REVIEW

**Comment by mrpatrickwest** [comment]
@zednis @am-e Current project update pull request

**Comment by mrpatrickwest** [comment]
The project I've been using, selecting the first project update

https://dco.tw.rpi.edu/vivo/display/n3053

**Comment by zednis** [comment]
Have we added any System.out or logging statements to this branch yet?

edit: adding @am-e

File: src/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddProjectUpdateGenerator.java [hunk 1]
```diff
@@ -315,4 +308,46 @@ public EditMode getEditMode(VitroRequest vreq) {
         return editMode;
     }
```
Comment (zednis) [inline src/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddProjectUpdateGenerator.java:341]: Would it be possible to put all of these queries into files that are loaded from the webapp resources?

**Comment by zednis** [comment]
I saw the note on AddProjectUpdatePreprocessor.preprocess( ) that the method does not seem to be called.  It appears that this method should be called by 

edu.cornell.mannlib.vitro.webapp.edit.n3editing.VTwo.ProcessRdfForm::applyEditSubmissionPreprocessors:

```
private void applyEditSubmissionPreprocessors(
            EditConfigurationVTwo configuration, MultiValueEditSubmission submission, VitroRequest vreq) {
        List<EditSubmissionVTwoPreprocessor> preprocessors = configuration.getEditSubmissionPreprocessors();
        if(preprocessors != null) {
            for(EditSubmissionVTwoPreprocessor p: preprocessors) {
                p.preprocess(submission, vreq);
            }
        }
    }
```

Let's add a (temporary) System.out statement in this method to make sure that our custom add project update preprocessor is in the preprocessors List.

File: src/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/AddProjectUpdatePreprocessor.java [hunk 2]
```diff
@@ -28,6 +28,8 @@ public AddProjectUpdatePreprocessor(EditConfigurationVTwo editConfig) {
 
     }
 
+/* FIXME: This doesn't seem to be called
+*/
     @Override
     public void preprocess(MultiValueEditSubmission inputSubmission, VitroRequest vreq) {
```
Comment (zednis) [inline src/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/preprocessors/AddProjectUpdatePreprocessor.java:26]: Let's add a System.out statement here and in `ProcessRdfForm::applyEditSubmissionPreprocessors` method to see when and if the preprocessor preprocess method is getting invoked.

**Comment by am-e** [comment]
I rebuilt on dco.tw but not sure how to run in debug mode.

**Comment by zednis** [comment]
@am-e Are you attempting to run the service in debug mode to set breakpoints and step through the code or to get the log.debug( ) statements?

edit:  to get the log.debug statements in the log you will need to find and edit the log4j.properties file.

**Comment by zednis** [comment]
guide on how to run VIVO in debug mode from Eclipse: https://wiki.duraspace.org/display/VIVO/Debugging+VIVO+in+Eclipse

**Comment by am-e** [comment]
@zednis  I'll look at the guide and try that out. I'd been placing System.println() statements and finding the output in catalina.out.

**Comment by zednis** [comment]
@am-e Don't worry about running the service in debug right now, but please replace your System.out statements with log statements (see my recent commit for an example) and commit the log statements to the branch.

**Comment by am-e** [comment]
Placed debug log statements throughout AddProjectUpdateGenerator.java, then rebuilt vivo. I set log4j.rootLogger=DEBUG, AllAppender in log4j.properties in my local instance and restarted Tomcat. Debug statements were now showing up in vivo.all.log but not the ones we placed. I'll try the same on dco.tw.

**Comment by mrpatrickwest** [comment]
when you click the pencil icon it now displays the current list of publications and instruments. Need to figure out now how to remove them if the user wants to. They can already remove it from the Project Update individual (click the label instead of the pencil) but we need to be able to remove them from the form. There are examples of this when creating concepts, just need to mimic it.

**Comment by mrpatrickwest** [comment]
https://dco.tw.rpi.edu/vivo/display/n3053

click the pencil next to PCW Test 01

still need to add the remove links, but almost there

**Comment by mrpatrickwest** [comment]
removing instruments now works. And there doesn't seem to be a problem now with adding instruments or publications. Need more testing, though.

**Comment by mrpatrickwest** [comment]
@zednis @am-e please test the project update form again. Everything should be working now.

**Comment by am-e** [comment]
All seems well with the editing on dco.tw.

**Comment by zednis** [comment]
:+1: I editing an existing and created a new project update and everything seemed to be working as expected.

edit: nm!  I found a new bug (see post below)

**Comment by zednis** [comment]
edit: I found a new bug!

This one is weird :-) 

On a project update with no selected publications add a duplicate instrument (I added "DCO Computer Cluster" twice), save, and reload the project update.

When the project update form reloads it now shows a listed publication with name `file:///var/lib/tomcat7/`

screen shot 2016-06-14 at 1 31 13 pm

**Comment by am-e** [comment]
There seem to be 2 issues here.
1. Why was a publication added at all?
2. The tomcat directory label which I think Patrick and I saw with pubs that had no labels.
Looking into both.

**Comment by zednis** [comment]
https://dco.tw.rpi.edu/vivo/display/n49812

The following ended up in the RDF:

```
<dco:associatedPublication rdf:resource="file:///var/lib/tomcat7/"/>
```

**Comment by mrpatrickwest** [comment]
Seems to be a problem in production as well. I've seen file:///usr/share/tomcat6, which is the tomcat path on deepcarbon production machine.

Also, same thing happens if you start with a publication. Select a publication, then select a duplicate and you get the same behavior for instruments.

I recommend that we create yet another ticket and work that one with this new issue. The branch, and the other one that is now fixed, can be shipped. Agreed???

**Comment by zednis** [comment]
@mrpatrickwest I agree, let's create a new ticket - especially since you have confirmed this behavior was not introduced in the changes related to this ticket.

I did not see any other issues with the project update edit form.  

:+1: to merging this branch.

**Pull request merged**